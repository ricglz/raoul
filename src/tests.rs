use std::{
    fs::{read_dir, ReadDir},
    io::Read,
};

use super::{parse, parse_ast, QuadrupleManager, VM};

fn parse_ast_has_error(filename: &str) {
    println!("Testing {:?}", filename);
    let program = std::fs::read_to_string(filename).expect(filename);
    let debug = false;
    let ast_response = parse(&program, debug);
    assert!(ast_response.is_ok());
    let ast = ast_response.unwrap();
    let res = parse_ast(ast, debug);
    assert!(res.is_err());
    insta::assert_debug_snapshot!(res.unwrap_err());
}

impl<R: Read> VM<R> {
    pub fn new_with_reader(quad_manager: &QuadrupleManager, debug: bool, reader: R) -> Self {
        VM::base_new(quad_manager, debug, Some(reader))
    }
}

fn parse_ast_is_ok(filename: &str) -> QuadrupleManager {
    println!("Testing {:?}", filename);
    let program = std::fs::read_to_string(filename).expect(filename);
    let debug = false;
    let ast_response = parse(&program, debug);
    assert!(ast_response.is_ok());
    let ast = ast_response.unwrap();
    let res = parse_ast(ast, debug);
    assert!(res.is_ok());
    let quad_manager = res.unwrap();
    insta::assert_debug_snapshot!(quad_manager);
    quad_manager
}

fn run_vm_is_error(filename: &str) {
    let quad_manager = parse_ast_is_ok(filename);
    let mut vm = VM::new_with_reader(&quad_manager, false, b"test".as_ref());
    let res = vm.run();
    assert!(res.is_err());
    insta::assert_display_snapshot!(res.unwrap_err());
    insta::assert_debug_snapshot!(vm.messages);
}

fn run_vm_is_ok(filename: &str) {
    let quad_manager = parse_ast_is_ok(filename);
    let mut vm = VM::new_with_reader(&quad_manager, false, b"test".as_ref());
    let res = vm.run();
    assert!(res.is_ok());
    insta::assert_debug_snapshot!(vm.messages);
}

fn expect_paths<F>(paths: ReadDir, mut f: F)
where
    F: FnMut(&str),
{
    for path in paths {
        let file_path = path.expect("File must exist").path();
        let file = file_path.to_str().unwrap();
        f(file);
    }
}

#[test]
fn ast_parsing_invalid_files() {
    let paths = read_dir("examples/invalid/static").unwrap();
    expect_paths(paths, parse_ast_has_error);
}

#[test]
fn vm_running_invalid_files() {
    let paths = read_dir("examples/invalid/dynamic").unwrap();
    expect_paths(paths, run_vm_is_error);
}

#[test]
fn valid_files() {
    let paths = read_dir("examples/valid").unwrap();
    for path in paths {
        let file_path = path.expect("File must exist").path();
        let file = file_path.to_str().unwrap();
        if file == "examples/valid/complete.ra" {
            continue;
        }
        run_vm_is_ok(file);
    }
}
