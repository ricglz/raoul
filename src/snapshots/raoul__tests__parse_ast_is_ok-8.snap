---
source: src/tests.rs
expression: quad_manager
---
QuadrupleManager {
    function_name: "main",
    jump_list: [],
    missing_return: false,
    dir_func: DirFunc {
        functions: {
            "main": Function {
                args: [],
                local_addresses: AddressManager(2, 2, 0, 7),
                name: "main",
                return_type: VOID,
                temp_addresses: TempAddressManager({
                    STRING: [],
                    INT: [
                        2001,
                    ],
                    FLOAT: [
                        2250,
                    ],
                    BOOL: [
                        2750,
                    ],
                }),
                variables: {
                    "i": Variable {
                        address: 1000,
                        data_type: INT,
                        dimensions: (
                            None,
                            None,
                        ),
                        name: "i",
                    },
                    "f": Variable {
                        address: 1755,
                        data_type: BOOL,
                        dimensions: (
                            None,
                            None,
                        ),
                        name: "f",
                    },
                    "a": Variable {
                        address: 1750,
                        data_type: BOOL,
                        dimensions: (
                            None,
                            None,
                        ),
                        name: "a",
                    },
                    "e": Variable {
                        address: 1754,
                        data_type: BOOL,
                        dimensions: (
                            None,
                            None,
                        ),
                        name: "e",
                    },
                    "j": Variable {
                        address: 1001,
                        data_type: INT,
                        dimensions: (
                            None,
                            None,
                        ),
                        name: "j",
                    },
                    "c": Variable {
                        address: 1752,
                        data_type: BOOL,
                        dimensions: (
                            None,
                            None,
                        ),
                        name: "c",
                    },
                    "h": Variable {
                        address: 1251,
                        data_type: FLOAT,
                        dimensions: (
                            None,
                            None,
                        ),
                        name: "h",
                    },
                    "d": Variable {
                        address: 1753,
                        data_type: BOOL,
                        dimensions: (
                            None,
                            None,
                        ),
                        name: "d",
                    },
                    "b": Variable {
                        address: 1751,
                        data_type: BOOL,
                        dimensions: (
                            None,
                            None,
                        ),
                        name: "b",
                    },
                    "k": Variable {
                        address: 1756,
                        data_type: BOOL,
                        dimensions: (
                            None,
                            None,
                        ),
                        name: "k",
                    },
                    "g": Variable {
                        address: 1250,
                        data_type: FLOAT,
                        dimensions: (
                            None,
                            None,
                        ),
                        name: "g",
                    },
                },
                first_quad: 0,
            },
        },
        global_fn: GlobalScope {
            addresses: AddressManager(0, 0, 0, 0),
            variables: {},
        },
    },
    memory: ConstantMemory({STRING: [1, 2.1], INT: [1, 2, 3, 4], FLOAT: [2], BOOL: [true, false]}),
    pointer_memory: PointerMemory {
        counter: 4000,
        pointers: {},
    },
    quad_list: [
        Goto       -     -     1    ,
        Or         3750  3751  2750 ,
        Assignment 2750  -     1750 ,
        And        3000  3750  2750 ,
        Assignment 2750  -     1751 ,
        Lte        3000  3001  2750 ,
        Assignment 2750  -     1752 ,
        Lt         3500  3001  2750 ,
        Assignment 2750  -     1753 ,
        Gte        3001  3000  2750 ,
        Assignment 2750  -     1754 ,
        Gt         3001  3000  2750 ,
        Assignment 2750  -     1755 ,
        Eq         3002  3002  2750 ,
        Assignment 2750  -     1754 ,
        Ne         3002  3001  2750 ,
        Assignment 2750  -     1755 ,
        Sum        3501  3250  2250 ,
        Assignment 2250  -     1250 ,
        Minus      3001  3250  2250 ,
        Assignment 2250  -     1251 ,
        Times      3003  3003  2000 ,
        Assignment 2000  -     1000 ,
        Div        3003  3003  2001 ,
        Assignment 2001  -     1001 ,
        Not        3751  -     2750 ,
        Assignment 2750  -     1756 ,
        Print      1750  -     -    ,
        Print      1751  -     -    ,
        Print      1752  -     -    ,
        Print      1753  -     -    ,
        Print      1754  -     -    ,
        Print      1755  -     -    ,
        Print      1250  -     -    ,
        Print      1251  -     -    ,
        Print      1000  -     -    ,
        Print      1001  -     -    ,
        Print      1756  -     -    ,
        PrintNl    -     -     -    ,
        End        -     -     -    ,
    ],
}
